### ISSUE TITLE
Improved Neural Arithmetic Logic Modules - layer to learn +-*/ operations in transparent way

### ISSUE THREAD

**Issue by FilipKubackiSoton** [tensorflow/addons#2768]
**Describe the feature and the current behavior/state.**
NALU (Neural Arithmetic Logic Modules) is the layer to learn addition, subtraction, multiplication, and division in a fully explainable/transparent way. Due to the mathematical nature of the layer, inference works well on OOD (out of distribution) inputs.
**Relevant information**
- Are you willing to contribute it (yes/no): yes
*If you wish to contribute, then read the requirements for new contributions in [`CONTRIBUTING.md`](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#requirements-for-new-contributions-to-the-repository)*
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): [NALU](https://arxiv.org/abs/1808.00508)
- Does the relavent academic paper exceed 50 citations? (yes/no): yes
- Is there already an implementation in another framework? (if so, where):  [inalu](https://github.com/daschloer/inalu)
- Was it part of tf.contrib? (if so, where):

**Which API type would this fall under (layer, metric, optimizer, etc.)**
layer
**Who will benefit with this feature?**
Everyone
**Any other info.**
To increase the robustness of NALU I can incorporate [iNALU](https://www.frontiersin.org/articles/10.3389/frai.2020.00071/fullhttps://papers.nips.cc/paper/2018/hash/0e64a7b00c83e3d22ce6b3acf2c582b6-Abstract.html) improvements using the following arguments: 
`gate_as_vector (bool)`: If True use gate as vector, otherwise like scalar
`clipping (float)`: Clipping value for weights if None do not use clipping
`force_operation (["add", "mul", None])`: force final result to belong to add/subtract, multiply/divide, or both. 
`input_gate_dependance (bool)`: If True make gate input dependent; otherwise independent.

**Comment by FilipKubackiSoton** [comment]
I realized that implementing only iNALU is not flexible enough. Therefore, I would like to close this issue due to the citation threshold. I want to raise a new issue related to NALU itself with configurable parameters that can utilize iNALU improvements. 

Community, can I do this without causing too much disruption, or is it better to modify this thread?

**Comment by AakashKumarNain** [comment]
Please use a single issue for tracking related to one topic

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
